app/uk/gov/hmrc/gform/service/RepeatingComponentService.scala

```diff
@@ -22,19 +22,149 @@ import play.api.Logger
 import play.api.libs.json.{ JsValue, Json }
 import uk.gov.hmrc.gform.connectors.SessionCacheConnector
 import uk.gov.hmrc.gform.gformbackend.model.FormTemplate
-import uk.gov.hmrc.gform.models.components.{ FieldId, FieldValue, Group }
+import uk.gov.hmrc.gform.models.components.{ FieldId, FieldValue, Group, _ }
+import uk.gov.hmrc.gform.models.{ Section, TextExpression, VariableInContext }
 import uk.gov.hmrc.http.cache.client.CacheMap
 import uk.gov.hmrc.play.http.HeaderCarrier
 
 import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.duration._
 import scala.concurrent.{ Await, Future }
+import scala.util.{ Success, Try }
 
 @Singleton
 class RepeatingComponentService @Inject() (val sessionCache: SessionCacheConnector) {
 
-  def getAllSections(formTemplate: FormTemplate)(implicit hc: HeaderCarrier) = {
-    Future.successful(formTemplate.sections)
+  def getAllSections(formTemplate: FormTemplate, data: Map[FieldId, Seq[String]])(implicit hc: HeaderCarrier) = {
+    sessionCache.fetch().map { maybeCacheMap =>
+      val cacheMap = maybeCacheMap.getOrElse(CacheMap("Empty", Map.empty))
+      formTemplate.sections.flatMap { section =>
+        if (isRepeatingSection(section)) {
+          generateDynamicSections(section, formTemplate, data, cacheMap)
+        } else {
+          List(section)
+        }
+      }
+    }
+  }
+
+  private def isRepeatingSection(section: Section) = section.repeatsMax.isDefined && section.fieldToTrack.isDefined
+
+  private def generateDynamicSections(section: Section, formTemplate: FormTemplate, data: Map[FieldId, Seq[String]],
+    cacheMap: CacheMap)(implicit hc: HeaderCarrier): List[Section] = {
+    val max = evaluateExpression(section.repeatsMax.get.expr, formTemplate, data)
+    val minRequested = evaluateExpression(section.repeatsMin.getOrElse(TextExpression(Constant("1"))).expr, formTemplate, data)
+    val min = if (minRequested <= 0) 1 else minRequested
+    val (maybeGroupField, requestedCount) = getRequestedCount(section.fieldToTrack.get, formTemplate, data, cacheMap)
+    val count = if (requestedCount >= min && requestedCount <= max) {
```

Comment:
maybe `scala.Predef.require(...)` would do it more readable. And we will get notified if someone passes wrong args here?